### DISCUSSION TITLE
Friend Registry Pointer broken?

### DISCUSSION THREAD

**Discussion by Vlhss** [YimMenu/YimMenu#467] [Q&A] [answered]
Going through the friend registry pointer returns empty friend's data while even sometimes some have random names or values like RID: `330`, Name: `T?` or `empty` etc. The only correct object it fetches is with index 0.
```
for (int i = 0; i < player_count; i++)
				player_button(*friend_registry->m_friends_list[i].m_data, i);
```
![](https://user-images.githubusercontent.com/106196444/194097497-5e5e98cf-a99c-4a2f-a382-1772b4910095.png)
Is the pointer perhaps broken or is there a patch that needs to be applied for the pointer to work correctly? I have no idea how it works or why it breaks.

**Comment by Yimura** [comment, marked as answer]
Wait what exactly are you doing here? This is how it's implemented in YimMenu:
```cpp
// FriendRegistry
main_batch.add("FR", "3B 0D ? ? ? ? 73 17", [this](memory::handle ptr)
{
	m_friend_registry = ptr.add(2).rip().as<FriendRegistry*>();

	LOG(G3LOG_DEBUG) << "FriendRegistry => [" << HEX_TO_UPPER(m_friend_registry) << "]";
});
```

And this is the example in the friends service:
https://github.com/YimMenu/YimMenu/blob/d01afda8d360baa527099d8952eb745cf75d5dee/BigBaseV2/src/services/friends/friends_service.cpp#L16-L26

To access it the following code should be correct:
```cpp
// please check the actual classes and definitions what you did originally is completely wrong
for (int i = 0; i < player_count; i++)
	player_button(friend_registry->m_friends_list.m_data[i], i);
```

I also checked in IDA and the signature still lands at the same variable.

Please also take a look at the actual classes:
```cpp
class FriendInfo
{
public:
	char m_name[20]; //0x0000
	char pad_0014[36]; //0x0014
	uint64_t m_rockstar_id; //0x0038
	uint8_t unk; //0x0040
	char pad_0041[3]; //0x0041
	uint32_t m_friend_state; //0x0044
	char pad_0048[304]; //0x0048
	uint32_t m_is_joinable; //0x0178
	char pad_017C[4]; //0x017C
}; //Size: 0x0180

class FriendList
{
public:
	class FriendInfo m_data[250]; //0x0000
}; //Size: 0x17700

#pragma pack(push, 4)
class FriendRegistry
{
public:
	uint32_t m_friend_count; //0x0000
	char pad_0004[8]; //0x0004
	class FriendList *m_friends_list; //0x000C
}; //Size: 0x0014
```

**Reply by Vlhss** [reply]
Yours implementation of the Friend Registry sig is exactly the same as mine.

As you wrote the following code should indeed be correct but for some reason it returns an object that has no info. Thus the no name green button as you can see on the picture I attached.

The friend_count is correct tho.

![image](https://user-images.githubusercontent.com/106196444/194372475-1604f988-6d5b-405c-9ac5-3ef17f058e90.png)
![image](https://user-images.githubusercontent.com/106196444/194372805-e4b761c5-0aa9-41f2-845c-b23dc37ded17.png)

**Reply by Vlhss** [reply]
Also I don't get how the original way I did it is wrong. Your friend service does the exact same thing.

`*friend_registry->m_friends_list[i].m_data`. Again, this line returns something random for the friend index of 1-friend_count apart from index 0;

**Reply by Yimura** [reply]
No it doesn't, I even linked to the friends service code above.

```cpp
g_pointers->m_friend_registry->m_friends_list->m_data[i]
```

**Reply by Yimura** [reply]
Please check the order of C++ operators, dereference operator happens as last.

1. Index of m_friends_list[i]
2. m_data (+0x0)
3. Dereference

Your code ends up somewhere completely differently than the code in YimMenu.

**Reply by Vlhss** [reply]
Yeah you are right, I completely forgot to give an index to the m_Data array I see it now.
